src/split_reduce.cpp

```diff
@@ -0,0 +1,208 @@
+            MIGRAPHX_THROW("should have one submodule.");
+        const auto* sm = mods.front();
+        if(sm->get_output_shapes().size() != 1)
+            MIGRAPHX_THROW("Only one output supported");
+        auto names = sm->get_parameter_names();
+        check_shapes{inputs, *this}.has(names.size()).same_ndims();
+        std::sort(names.begin(), names.end());
+        auto shapes = sm->get_parameter_shapes();
+        // Check dimension matches for each input
+        if(not equal(names, inputs, [&](const auto& name, const auto& input) {
+               return shapes.at(name).lens() == input.lens();
+           }))
+            MIGRAPHX_THROW("Dimenstion does not match the submodule.");
+        const auto& s = inputs.at(0);
+        auto lens     = s.lens();
+        if(lens != sm->get_output_shapes().front().lens())
+        {
+            for(const auto& axis : axes)
+            {
+                lens[axis] = 1;
+            }
+        }
+
+        return shape::from_permutation(
+            sm->get_output_shapes().front().type(), lens, find_permutation(inputs));
+    }
+
+    std::string name() const { return "split_fused_reduce"; }
+};
+MIGRAPHX_REGISTER_OP(split_fused_reduce);
+
+static bool is_reduce(const instruction& ins) { return contains(ins.name(), "reduce"); }
+
+static std::vector<instruction_ref> find_split(const_module_ref rm)
+{
+    std::vector<instruction_ref> result;
+    auto reduce_ins = std::find_if(rm->begin(), rm->end(), &is_reduce);
+    if(reduce_ins == rm->end())
+        return result;
+    // Bail if there is more than one reduce for now
+    if(std::any_of(std::next(reduce_ins), rm->end(), &is_reduce))
+        return result;
+    // Only handle reduce_sum for now
+    if(reduce_ins->name() != "reduce_sum")
+        return result;
+    result.push_back(reduce_ins);
+    // TODO: Find instructions that are used again in the module
+    return result;
+}
+
+static std::vector<instruction_ref> get_alive(const_module_ref rm,
+                                              const std::vector<instruction_ref>& splits)
+{
+    std::vector<instruction_ref> result;
+    bool stop = false;
+    liveness(*rm, [&](auto ins, const auto& live_set) {
+        if(stop)
+            return;
+        if(not contains(splits, ins))
+            return;
+        std::copy_if(live_set.begin(),
+                     live_set.end(),
+                     std::back_inserter(result),
+                     [](instruction_ref live) { return live->name() != "@param"; });
+        stop = true;
+    });
+    return result;
+}
+
+static std::string assign_op(const std::vector<instruction_ref>& splits)
+{
+    static std::unordered_map<std::string, std::string> m = {
+        {"reduce_sum", "assign_add"},
+        {"reduce_mean", "assign_add"},
+        {"reduce_prod", "assign_mul"},
+        {"reduce_max", "assign_max"},
+        {"reduce_min", "assign_min"},
+    };
+    return m.at(splits.front()->name());
```

Comment:
Doesn't it need to return vector as the output as the inputs are also vector ?

Reply:
Yes but for now we dont support multiple outputs. This will switch to a vector when we support the multiple outputs.